### PR TITLE
Implement telemetry for acquisition

### PR DESCRIPTION
**Pull request by chuckries**
1. Create new PackageError class that captures package info to be thrown
from all package related errors.

2. Send telemetry event with relevant platform/package information.

### CODE REVIEW

**Comment by chuckries** [comment]
@gregg-miskelly @DustinCampbell @wesrupert

File: src/omnisharp/server.ts [hunk 1]
```diff
@@ -319,7 +319,8 @@ export abstract class OmnisharpServer {
             cleanupPromise = new Promise<void>((resolve, reject) => {
                 const killer = exec(`taskkill /F /T /PID ${this._serverProcess.pid}`, (err, stdout, stderr) => {
                     if (err) {
-                        return reject(err);
+                        reject(err);
+                        return;
```
Comment (DustinCampbell) [inline src/omnisharp/server.ts]: Any reason you're changing these? It's perfectly legal to return a void-returning function.

File: src/packages.ts [hunk 2]
```diff
@@ -154,22 +162,21 @@ function downloadFile(urlString: string, pkg: Package, logger: Logger, status: S
...
         }
 
         let request = https.request(options, response => {
             if (response.statusCode === 301 || response.statusCode === 302) {
                 // Redirect - download from new location
-                return downloadFile(response.headers.location, pkg, logger, status)
-                    .then(
-                        () => resolve(),
-                        err => reject(err));
+                resolve(downloadFile(response.headers.location, pkg, logger, status));
```
Comment (DustinCampbell) [inline src/packages.ts]: Why this change? We're now calling resolve even if the `downloadFile` promise rejects.

File: src/packages.ts [hunk 3]
```diff
@@ -204,15 +211,15 @@ function downloadFile(urlString: string, pkg: Package, logger: Logger, status: S
             });
 
             response.on('error', err => {
-                reject(err)
+                reject(new PackageError('Reponse error:' + err.code || '', pkg, err));
```
Comment (DustinCampbell) [inline src/packages.ts]: missed space? Maybe just use a template literal?

File: src/omnisharp/server.ts (hunk 1 above)
Comment (DustinCampbell) [inline src/omnisharp/server.ts]: FWIW, I personally dislike this being on two lines when it could easily be on one. So, unless there's a real semantic difference I'm unaware of, I'd prefer not making these changes.

File: src/packages.ts [hunk 4]
```diff
@@ -230,12 +237,14 @@ function installPackage(pkg: Package, logger: Logger, status?: Status): Promise<
...
         if (!pkg.tmpFile || pkg.tmpFile.fd == 0) {
-            return reject(new Error('Downloaded file unavailable'));
+            reject(new PackageError('Downloaded file unavailable', pkg));
+            return;
         }
 
         yauzl.fromFd(pkg.tmpFile.fd, { lazyEntries: true }, (err, zipFile) => {
             if (err) {
-                return reject(err);
+                reject(new PackageError('Immediate zip file error:', pkg, err));
```
Comment (DustinCampbell) [inline src/packages.ts]: Unnecessary `:`?

File: src/main.ts [hunk 5]
```diff
@@ -101,18 +104,48 @@ function installRuntimeDependencies(extension: vscode.Extension<any>, logger: Lo
...
+            telemetryProps['installStage'] = installationStage;
+            telemetryProps['platform.architecture'] = platformInfo.architecture;
+            telemetryProps['platform.platform'] = platformInfo.platform;
+            if (platformInfo.distribution) { telemetryProps['platform.distribution'] = platformInfo.distribution.toString(); }
+            telemetryProps['platform.runtimeId'] = platformInfo.runtimeId;
+
+            _reporter.sendTelemetryEvent(
+                "Acquisition",
+                telemetryProps
+                );
```
Comment (DustinCampbell) [inline src/main.ts]: Why is this split across four lines?

File: src/main.ts [hunk 6]
```diff
@@ -101,18 +104,48 @@ function installRuntimeDependencies(extension: vscode.Extension<any>, logger: Lo
...
+            }
+
             logger.appendLine(`Failed at stage: ${installationStage}`);
             logger.appendLine(errorMessage);
         })
         .then(() => {
+            telemetryProps['installStage'] = installationStage;
+            telemetryProps['platform.architecture'] = platformInfo.architecture;
+            telemetryProps['platform.platform'] = platformInfo.platform;
+            if (platformInfo.distribution) { telemetryProps['platform.distribution'] = platformInfo.distribution.toString(); }
```
Comment (DustinCampbell) [inline src/main.ts]: Total nit: This should probably be on three lines.

File: src/main.ts [hunk 7]
```diff
@@ -101,18 +104,48 @@ function installRuntimeDependencies(extension: vscode.Extension<any>, logger: Lo
             installationStage = 'touchLockFile';
             return util.touchInstallFile(util.InstallFileType.Lock);
         })
+        .then(() => {
+            installationStage = 'completeSuccess';
+        })
         .catch(error => {
-            errorMessage = error.toString();
+            if (error instanceof PackageError) {
+                // we can log the message in a PackageError to telemetry as we do not put PII in PackageError messages
```
Comment (DustinCampbell) [inline src/main.ts]: Good idea.

File: src/main.ts [hunk 8]
```diff
@@ -12,18 +12,19 @@ import * as coreclrdebug from './coreclr-debug/activate';
 import * as OmniSharp from './omnisharp/extension';
 import * as util from './common';
 import { Logger } from './logger';
-import { PackageManager, Status } from './packages';
+import { PackageManager, Status, PackageError } from './packages';
 import { PlatformInformation } from './platform';
 
 let _channel: vscode.OutputChannel = null;
+let _reporter: TelemetryReporter = null;
```
Comment (DustinCampbell) [inline src/main.ts]: Feels inconsistent to make this a global variable while we thread `logger` through. Should we just thread this through the functions with `logger`?

Comment (chuckries) [inline src/main.ts]: Probably.

File: src/omnisharp/server.ts (hunk 1 above)
Comment (chuckries) [inline src/omnisharp/server.ts]: I wasn't familiar with this pattern. I can change them back.

File: src/packages.ts (hunk 2 above)
Comment (DustinCampbell) [inline src/packages.ts]: Note: This is the change I'm most interested. Everything else is nits.

Comment (chuckries) [inline src/packages.ts]: We are resolving with a promise that may reject, simply continuing the promise chain. There's no need to try to capture errors here as downloadFile captures error in all of it's non recursive code paths. 
https://jakearchibald.com/2014/resolve-not-opposite-of-reject/ 

It looks like both this way and the way this was originally written leaves us open to infinite redirect...

File: src/main.ts (hunk 5 above)
Comment (chuckries) [inline src/main.ts]: The telemetryProps object was initially created as a block here. I'll compress.

File: src/packages.ts (hunk 2 above)
Comment (DustinCampbell) [inline src/packages.ts]: Ah yes, you're right.

**Comment by DustinCampbell** [comment]
I don't have any concerns with this change. It'd be nice if some of the nits were cleaned up though.

**Comment by chuckries** [comment]
I appreciate the nits, my javascript/typescript conventions are still rusty.

**Comment by chuckries** [comment]
Fixed up the nits, merging.

**Comment by gregg-miskelly** [comment]
@chuckries in the definition of PackageError, I would suggest adding a comment that 'message' should not contain PII

File: src/packages.ts [hunk 9]
```diff
@@ -154,22 +161,19 @@ function downloadFile(urlString: string, pkg: Package, logger: Logger, status: S
...
         }
 
         let request = https.request(options, response => {
             if (response.statusCode === 301 || response.statusCode === 302) {
                 // Redirect - download from new location
-                return downloadFile(response.headers.location, pkg, logger, status)
-                    .then(
-                        () => resolve(),
-                        err => reject(err));
+                resolve(downloadFile(response.headers.location, pkg, logger, status));
```
Comment (gregg-miskelly) [inline src/packages.ts]: @chuckries You lost the 'return' here. Is that intentional?

Comment (chuckries) [inline src/packages.ts]: good catch, see #893
